### PR TITLE
0.7 only - drop 0.6

### PR DESCRIPTION
**Pull request by SimonDanisch**

### CODE REVIEW

**Comment by maleadt** [comment]
Rebased and bumped CI. I assume this supersedes #108?

**Comment by SimonDanisch** [comment]
yes! Just wanted to wait with closing until femtocleaner comes around!

**Comment by maleadt** [comment]
https://github.com/JuliaGPU/GPUArrays.jl/pull/127/commits/840b2cc4cc4dc67ddd6c1fc2a1d2bb908a59d3d5

**Comment by MikeInnes** [comment]
I dropped the femtocleaner commit, only the `Complex` fix was correct and it broke package loading.

File: appveyor.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 environment:
   matrix:
-  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.6/julia-0.6-latest-win64.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.7/julia-0.7.0-alpha-win64.exe"
```
Comment (vchuravy) [inline appveyor.yml]: this should be `julia-0.7-latest-win64.exe`

File: src/broadcast.jl [hunk 2]
```diff
@@ -1,158 +1,110 @@
...
+        f(state, cartesian, A, args...)
+        return
+    end
 end
-function broadcast_t(f::Any, ::Type{Any}, ::Any, ::Any, A::GPUArrays.GPUArray, args::Vararg{Any,N}) where N
-    error("Return type couldn't be inferred for broadcast. Func: $f, $(typeof(A)), $args")
+
+# copyto! overloads
+@inline function copyto!(dest::GPUArray, B::GPUBroadcast)
+    flat = flatten(B); as = flat.args; f = flat.f
```
Comment (vchuravy) [inline src/broadcast.jl]: You don't need to flatten the Broadcast as long as you can convert both `Broadcasted` and `Extruded` to GPU representations:

```
function CUDAnative.cudaconvert(bc::Broadcast.Broadcasted{Style}) where Style
    Broadcast.Broadcasted{Style}(bc.f, map(CUDAnative.cudaconvert, bc.args), bc.axes)
end

function CUDAnative.cudaconvert(ex::Broadcast.Extruded)
    Broadcast.Extruded(CUDAnative.cudaconvert(ex.x), ex.keeps, ex.defaults)
end
```

which simplifies broadcast to:

```
function _cuda_copyto_kernel!(dest::CuDeviceArray, bc::Broadcast.Broadcasted{Nothing})
    shape = axes(dest)
    let I = @cuda_index(shape)
        @inbounds dest[I] = bc[I]
    end
end
```

**Comment by vchuravy** [comment]
x-ref: #129

**Comment by SimonDanisch** [comment]
Do we wait for the todos in https://github.com/JuliaGPU/CuArrays.jl/pull/99 ? Or do we merge and open new PR's for specific issues?

**Comment by MikeInnes** [comment]
I'd personally be fine with tagging a buggy beta release of both packages (which is strictly better than the current situation) and working from there.

**Comment by vchuravy** [comment]
I think getting it on master is fine, but I would be hesitant to put out a release until at least one of the GPU backends works.

**Comment by SimonDanisch** [comment]
That seems reasonable!

**Pull request merged**